src/dg/dg.cpp

```diff
@@ -2095,7 +2095,7 @@ void DGBase<dim,real,MeshType>::allocate_system (
 
     dof_handler.distribute_dofs(fe_collection);
     //This Cuthill_McKee renumbering for dof_handlr uses a lot of memory in 3D, is there another way?
-    dealii::DoFRenumbering::Cuthill_McKee(dof_handler,true);
+    if(all_parameters->do_renumber_dofs) dealii::DoFRenumbering::Cuthill_McKee(dof_handler,true);
```

Comment:
Might be a question for @dougshidong - what is the purpose of this renumbering and why is it such a memory bottleneck?

Reply:
Big paragraph ahead. Simply because I'm procrastinating on work I should actually be doing.

### What it doesn't do

Reordering does *not* change inherently changes the conditioning of your matrix. Your eigenvalue should stay the same, but your eigenvectors will be switched around. It's the same problem, from a different angle. [1, 3; 2 9] has the same eigenvalues as [9 2; 3 1]

### What it does
What reodering does offer is a lower fill-in when performing ILU preconditioning.

### LU Factorization
#### Sidebar 1
ILU aims to approximate P~A^{-1} such that solving P*A x = P*b is "easy". ILU approximates A^{-1} by doing the usual LU factorization. However, it will only store values v_{i,j} in L and U if the value of A^k is non-zero (k being ILU(k+1)). The new values are called fill-in. It essentially tries as keep a semblance of the sparsity pattern of A. 
#### Sidebar 2
The bandwidth of a matrix is defined by the maximum distance an entry is from the diagonal. Aka, how tightly diagonal it is. The LU algorithm is in nature very sequential (unless you ask Aditya Kashi), such that the the operation you do on one row depends on the previous operation. Now, if you can imagine how keeping a low bandwidth propagates as much information as possible while not creating too many non-zeros. Great [video](https://www.youtube.com/watch?v=ymIvps7pgRk) that gives a feel for "information propagation" based on numbering.
#### Impact
Now, if A has a "random" sparsity pattern, the LU factorization of that matrix will not necessarily correspond to the sparsity pattern of the original matrix. Therefore, the incomplete factorization will likely drop many entries that would have otherwise been important. However, when reordering DoFs using Cuthill_McKee, we somewhat reduce the *bandwidth* of the matrix. Notice the many -ly and uncertain words above; this is because a lot of this is based on empirical tests and heuristics.


### Actually trying to answer your question instead of babbling

> this reorders the cells such that we have a block diagonal matrix

Cuthill_Mckee does not form block diagonals. It tries to minimize the bandwidth of the matrix. This is a heuristic to minimize graph bandwidth, it does not actually minimize it.

> which is essential for the conditioning of the linear solver

Yeah, kind of. It is "essential" specifically for ILU in CFD. However, it can also be used for LUSGS algorithms like in the video above that requires no matrix to be stored. Cuthill-McKee specifically is just a heuristic.

> Looping over all of the dofs is understandably memory-intensive.

Kind of false. Looping over the dofs take no memory. It's what you do inside of it. My understanding is that some variation of breadth-first search is done on the adjacency matrix. The adjacency matrix (likely implemented as adjacency list that will store as many entries as non-zeroes) will then require memory proportional to the actual sparse matrix.

### Reordering

Give this [link](https://www.dealii.org/current/doxygen/deal.II/namespaceDoFRenumbering.html) a read.

Reordering can be important for so much more than ILU. Sometimes, you may want to separate your equations, like density, momentum, and energy because you can use a specific preconditioner on momentum, and another one on density and energy. In this scenario, a component-wise reordering would be used. Or maybe you rely on the block-property of DG cells, because you want to apply a transformation with some underlying physics. Then you would use cell-wise numbering. You can even use some fancy reordering scheme in a way that numbers "emanate" from airfoil to farfield if you believe that physical information propagates that way. Or maybe you just want things to be close to each other on your computer such that accessing them is faster due to less cache misses.

### Takeaway

Renumbering has 0 impact on the result of explicit time-stepping because there is no dependency on ordering. However, it may still impact its memory-access efficiency.

Reordering is very important for any type of implicit schemes because it affects the order in which information is propagated from cell to cell.

Reply:
Also, we should probably keep reordering on for any implicit schemes. There is little to no chance most iterative implicit schemes converge for a random ordering.

Reply:
Thank you for the detailed explanation! I agree that my explanation was vague (and, as you pointed out, a bit incorrect in places!). I will read through the links you provided, it's always best to understand what's going on in the code.